staging/src/k8s.io/client-go/tools/remotecommand/websocket.go

```diff
@@ -0,0 +1,342 @@
+		streams: map[byte]*stream{},
+	}
+
+	go ws.sendPings(PingPeriod) // start heartbeat
+
+	return &ws
+}
+
+func (c *wsStreamCreator) getStream(id byte) *stream {
+	c.streamsMu.Lock()
+	defer c.streamsMu.Unlock()
+	return c.streams[id]
+}
+
+func (c *wsStreamCreator) setStream(id byte, s *stream) {
+	c.streamsMu.Lock()
+	defer c.streamsMu.Unlock()
+	c.streams[id] = s
+}
+
+// CreateStream uses id from passed headers to create a stream over "c.conn" connection.
+// Returns a Stream structure or nil and an error if one occurred.
+func (c *wsStreamCreator) CreateStream(headers http.Header) (httpstream.Stream, error) {
+	streamType := headers.Get(v1.StreamType)
+	id, ok := streamType2streamID[streamType]
+	if !ok {
+		return nil, fmt.Errorf("unknown stream type: %s", streamType)
+	}
+	if s := c.getStream(id); s != nil {
+		return nil, fmt.Errorf("duplicate stream for type %s", streamType)
+	}
+	reader, writer := io.Pipe()
+	s := &stream{
+		headers:   headers,
+		readPipe:  reader,
+		writePipe: writer,
+		conn:      c.conn,
+		connMu:    &c.connMu,
+		id:        id,
+	}
+	c.setStream(id, s)
+	return s, nil
+}
+
+// sendPings starts heartbeat sending a "ping" from the endpoint every "period".
+func (c *wsStreamCreator) sendPings(period time.Duration) {
+	c.connMu.Lock()
+	closedCh := make(chan struct{})
+	closeHandler := c.conn.CloseHandler()
+	c.conn.SetCloseHandler(func(code int, text string) error {
+		close(closedCh)
+		return closeHandler(code, text)
+	})
+	c.connMu.Unlock()
+	t := time.NewTicker(period)
+	defer t.Stop()
+	for {
+		select {
+		case <-closedCh:
+			return
+		case <-t.C:
+			c.connMu.Lock()
+			if err := c.conn.WriteControl(gwebsocket.PingMessage, nil, time.Now().Add(period)); err != nil {
+				klog.V(7).Infof("Websocket Ping failed: %v", err)
+				// Continue, in case this is a transient failure.
+				// c.conn.CloseChan above will tell us when the connection is
+				// actually closed.
+			} else {
+				klog.V(7).Infof("Websocket Ping succeeeded")
+			}
+			c.connMu.Unlock()
+		}
+	}
+}
+
+func (c *wsStreamCreator) run() {
+	readBuffer := make([]byte, remotecommand.WebSocketBufferSize)
+	// Set up handler for ping/pong heartbeat.
+	if err := c.conn.SetReadDeadline(time.Now().Add(pingReadDeadline)); err != nil {
+		klog.V(7).Infof("Websocket setting read deadline failed %v", err)
+	}
+	c.conn.SetPongHandler(func(string) error {
+		err := c.conn.SetReadDeadline(time.Now().Add(pingReadDeadline))
+		if err != nil {
+			klog.V(7).Infof("Websocket setting read deadline failed %v", err)
+		}
+		return nil
+	})
+	for {
+		messageType, r, err := c.conn.NextReader()
+		if err != nil {
+			websocketErr, ok := err.(*gwebsocket.CloseError)
+			if ok && websocketErr.Code == gwebsocket.CloseNormalClosure {
+				err = nil // readers will get io.EOF as it's a normal closure
+			} else {
+				err = fmt.Errorf("next reader: %w", err)
+			}
+			c.closeAllStreamReaders(err)
+			return
+		}
+		if messageType != gwebsocket.BinaryMessage {
```

Comment:
I'm not sure about this, but it depends on what NextReader() is returning, so websocket defines opcodes in https://datatracker.ietf.org/doc/html/rfc6455#section-1.2

```
   Opcode:  4 bits

      Defines the interpretation of the "Payload data".  If an unknown
      opcode is received, the receiving endpoint MUST _Fail the
      WebSocket Connection_.  The following values are defined.

      *  %x0 denotes a continuation frame

      *  %x1 denotes a text frame

      *  %x2 denotes a binary frame

      *  %x3-7 are reserved for further non-control frames

      *  %x8 denotes a connection close

      *  %x9 denotes a ping

      *  %xA denotes a pong

      *  %xB-F are reserved for further control frames
```

what we want to process here are the `%x2 denotes a binary frame` but I'm not sure we close the connection on the other opcodes and just continue and let the server process them

Reply:
maybe `log` at debug level and `continue` ?

Reply:
According to the documentation for `conn.NextReader()`
```
// NextReader returns the next data message received from the peer. The                                                                                                                               
// returned messageType is either TextMessage or BinaryMessage.    
```
The `NextReader()` does not return Control frames. So the error here is saying that we've unexpectedly received a `TextFrame`. This should be an error, since this WebSocket client only streams binary data messages (even the base64-encoded messages are binary; and we do not even use the base64-encoded data messages in any of the versions of the remote command subprotocol). That said, I think it would be fine to log(debug) and continue, however. Please let me know what you think.

Reply:
did this discussion get resolved and is the resolution well-tested?

Reply:
I am confident this is a non-issue. Antonio left on vacation before we could resolve it.

Reply:
+1